---
title: "MultiNicheNet analysis: Integrated lung atlas analysis - correct for batch effects to infer differences between IPF and healthy subjects - wrapper function"
author: "Robin Browaeys"
date: "2023-06-06"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{MultiNicheNet analysis: Integrated lung atlas analysis - correct for batch effects to infer differences between IPF and healthy subjects - wrapper function}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<!-- github markdown built using 
rmarkdown::render("vignettes/batch_correction_analysis_LungAtlas.Rmd", output_format = "github_document")
-->

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  # comment = "#>",
  warning = FALSE,
  message = FALSE
)
```

In this vignette, you can learn how to perform a MultiNicheNet analysis comparing cell-cell communication between different groups/conditions of interest while correcting for batch effects. In this vignette, we start from one SingleCellExperiment object containing cells from both sender and receiver cell types and from different patients. This object was obtained after merging cells from different studies as described below.

A MultiNicheNet analysis can be performed if you have multi-sample, multi-group single-cell data. MultiNicheNet will look for cell-cell communication between the cell types in your data for each sample, and compare the cell-cell communication patterns between the groups of interest. Therefore, the absolute minimum of meta data you need to have, are following columns indicating for each cell: the **group**, **sample** and **cell type**.
 
As example expression data of interacting cells, we will here use merged scRNAseq data from four studies comparing healthy lungs to lungs from patients with idiopathic pulmonary fibrosis (IPF)) (Adams 2020, Reyfman 2019, Morse 2019, and Habermann 2020). Harmonized cell type annotations across the 4 different studies were obtained through Azimuth [Azimuth meta-analysis of human scRNA-seq datasets](https://cellxgene.cziscience.com/collections/2f75d249-1bec-459b-bf2b-b86221097ced).

We will here demonstrate how MultiNicheNet can exploit the flexibility of generalized linear models in the pseudobulk-edgeR framework to correct for batch effects, here the source study: Adams 2020, Reyfman 2019, Morse 2019, or Habermann 2020. We will apply MultiNicheNet to compare cell-cell interaction changes between IPF and healthy tissue. Note that the only required input for a batch-correcting MultiNicheNet analysis is a merged scRNA-seq object containing raw counts and harmonized cell type annotations.

In this vignette, we will first prepare the data and analysis parameters, and then perform the MultiNicheNet analysis. 

The different steps of the MultiNicheNet analysis are the following:

* 0. Preparation of the analysis: load packages, NicheNet LR network & ligand-target matrix, single-cell expression data, and define main settings of the MultiNicheNet analysis

* 1. Extract cell type abundance and expression information from receiver and sender cell types, and link this expression information for ligands of the sender cell types to the corresponding receptors of the receiver cell types 

* 2. Perform genome-wide differential expression analysis of receiver and sender cell types to define DE genes between the conditions of interest. Based on this analysis, we can define the logFC/p-value of ligands in senders and receptors in receivers, and define the set of affected target genes in the receiver.

* 3. Predict NicheNet ligand activities and NicheNet ligand-target links based on these differential expression results

* 4. Use the information collected above to prioritize all sender-ligand---receiver-receptor pairs.

* 5. Calculate correlation in expression between ligand-receptor pairs and their predicted target genes

In this vignette, we will demonstrate the use of a wrapper function to perform all these steps in one line of code. If you want to explore the different steps of MultiNicheNet one by one in more detail, you could check this other vignette: [basis_analysis_steps_MISC.md](basis_analysis_steps_MISC.md).

After the MultiNicheNet analysis is done, we will explore the output of the analysis with different ways of visualization. 

# Step 0: Preparation of the analysis: load packages, NicheNet LR network & ligand-target matrix, single-cell expression data

## Step 0.1: Load required packages and NicheNet ligand-receptor network and ligand-target matrix

```{r}
library(SingleCellExperiment)
library(dplyr)
library(ggplot2)
library(multinichenetr)
```

The Nichenet v2 networks and matrices for both mouse and human can be downloaded from Zenodo [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.7074291.svg)](https://doi.org/10.5281/zenodo.7074291). 

We will read these object in for human because our expression data is of human patients. 
Gene names are here made syntactically valid via `make.names()` to avoid the loss of genes (eg H2-M3) in downstream visualizations.

```{r}
organism = "human"
if(organism == "human"){
  lr_network = readRDS(url("https://zenodo.org/record/7074291/files/lr_network_human_21122021.rds"))
  lr_network = lr_network %>% dplyr::rename(ligand = from, receptor = to) %>% distinct(ligand, receptor) %>% mutate(ligand = make.names(ligand), receptor = make.names(receptor))
  ligand_target_matrix = readRDS(url("https://zenodo.org/record/7074291/files/ligand_target_matrix_nsga2r_final.rds"))
  colnames(ligand_target_matrix) = colnames(ligand_target_matrix) %>% make.names()
  rownames(ligand_target_matrix) = rownames(ligand_target_matrix) %>% make.names()
} else if(organism == "mouse"){
  lr_network = readRDS(url("https://zenodo.org/record/7074291/files/lr_network_mouse_21122021.rds"))
  lr_network = lr_network %>% dplyr::rename(ligand = from, receptor = to) %>% distinct(ligand, receptor) %>% mutate(ligand = make.names(ligand), receptor = make.names(receptor))
  ligand_target_matrix = readRDS(url("https://zenodo.org/record/7074291/files/ligand_target_matrix_nsga2r_final_mouse.rds"))
  colnames(ligand_target_matrix) = colnames(ligand_target_matrix) %>% make.names()
  rownames(ligand_target_matrix) = rownames(ligand_target_matrix) %>% make.names()
}
```

## Step 0.2: Read in SingleCellExperiment Objects 

In this vignette, sender and receiver cell types are in the same SingleCellExperiment object, which we will load here. In this vignette, we will load in a subset of the scRNAseq data of the MIS-C [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.8010790.svg)](https://doi.org/10.5281/zenodo.8010790). For the sake of demonstration, this subset only contains 4 cell types. 

If you start from a Seurat object, you can convert it easily to a SingleCellExperiment via `sce = Seurat::as.SingleCellExperiment(seurat_obj, assay = "RNA")`.

Because the NicheNet 2.0. networks are in the most recent version of the official gene symbols, we will make sure that the gene symbols used in the expression data are also updated (= converted from their "aliases" to official gene symbols). Afterwards, we will make them again syntactically valid. 

```{r}
sce = readRDS(url("https://zenodo.org/record/8010790/files/sce_subset_lung.rds"))
sce = alias_to_symbol_SCE(sce, "human") %>% makenames_SCE()
```

Now we will go further in defining the settings for the MultiNicheNet analysis

# Step 1:  Prepare the cell-cell communication analysis

### Define in which metadata columns we can find the **group**, **sample** and **cell type** IDs

In this case study, we want to study differences in cell-cell communication changes between two patient groups (IPF patients vs healthy subjects), while considering the source dataset cells were derived from. Patient groups are indicated in the following meta data column: `disease`, which has 2 different values: idiopathic.pulmonary.fibrosis and normal.

Cell type annotations are indicated in the `annotation.l1` column, and the sample is indicated by the `donor` column. 
If your cells are annotated in multiple hierarchical levels, we recommend using a high level in the hierarchy. This for 2 reasons: 1) MultiNicheNet focuses on differential expression and not differential abundance, and 2) there should be sufficient cells per sample-celltype combination.

If you would have batch effects or covariates you can correct for, you can define this here as well. Here, we want to correct for the source study, which is indicated in the following meta data column: `dataset_origin`, which has 4 different values: adams_2020, habermann_2020, morse_2019, reyfman_2019 and normal.

Important: for categorical covariates and batches, there should be at least one sample for every group-batch combination. If one of your groups/conditions lacks a certain level of your batch, you won't be able to correct for the batch effect because the model is then not able to distinguish batch from group/condition effects.

Important: the column names of group, sample, cell type, batches and covariates should be syntactically valid (`make.names`)

Important: All group, sample, cell type, batch and covariate names should be syntactically valid as well (`make.names`) (eg through `SummarizedExperiment::colData(sce)$ShortID = SummarizedExperiment::colData(sce)$ShortID %>% make.names()`)

```{r}
sample_id = "donor"
group_id = "disease"
celltype_id = "annotation.l1"
covariates = NA
batches =  "dataset_origin"
```

Important: for a MultiNicheNet analysis there is a difference between a covariate and batch in the following sense: covariates will just be included in the DE GLM model, whereas batches will be included in the DE GLM model AND expression values for downstream visualizations will be corrected for the batch effects. In this dataset, we want to take into account the dataset-of-origin and correct the expression values for downstream visualizations (as demonstrated later). Therefore we add dataset_origin as batch and not as covariate.

Sender and receiver cell types also need to be defined. Both are here all cell types in the dataset because we are interested in an All-vs-All analysis.

```{r}
senders_oi = SummarizedExperiment::colData(sce)[,celltype_id] %>% unique()
receivers_oi = SummarizedExperiment::colData(sce)[,celltype_id] %>% unique()
```

If the user wants it, it is possible to use only a subset of senders and receivers. Senders and receivers can be entirely different, but also overlapping, or the same. If you don't use all the cell types in your data, we recommend to continue with a subset of your data.

```{r}
sce = sce[, SummarizedExperiment::colData(sce)[,celltype_id] %in% c(senders_oi, receivers_oi)]
```


### Define the contrasts and covariates of interest for the DE analysis, and the minimal number of cells of a cell type that each sample should have to be considered for DE analysis of that cell type.

Since MultiNicheNet will infer group differences at the sample level for each cell type (currently via Muscat - pseudobulking + EdgeR), we need to have sufficient cells per sample of a cell type, and this for both groups. In the following analysis we will set this minimum number of cells per cell type per sample at 10 (recommended minimum). 

```{r}
min_cells = 10
```

For this analysis, we want to compare how cell-cell communication differs between IPF and normal lungs.

To do this comparison, we need to set the following contrasts:

```{r}
contrasts_oi = c("'idiopathic.pulmonary.fibrosis-normal','normal-idiopathic.pulmonary.fibrosis'")
```

__Very Important__ Note the format to indicate the contrasts! This formatting should be adhered to very strictly, and white spaces are not allowed!  Check `?get_DE_info` for explanation about how to define this well. The most important things are that: each contrast is surrounded by single quotation marks, contrasts are separated by a comma without any whitespace, and alle contrasts together are surrounded by double quotation marks. If you compare against two groups, you should divide by 2, if you compare against three groups, you should divide by 3 etcetera. See some of the other MultiNicheNet tutorials that demonstrate how to do this.

For downstream visualizations and linking contrasts to their main group, you need to run the following:

```{r}
contrast_tbl = tibble(contrast =
                        c("idiopathic.pulmonary.fibrosis-normal", "normal-idiopathic.pulmonary.fibrosis"),
                      group = c("idiopathic.pulmonary.fibrosis", "normal"))
```

### Define the parameters for the NicheNet ligand activity analysis 

Here, we need to define the thresholds that will be used to consider genes as differentially expressed or not (logFC, p-value, decision whether to use adjusted or normal p-value, minimum fraction of cells that should express a gene in at least one sample in a group, whether to use the normal p-values or empirical p-values). 

NicheNet ligand activity will then be calculated as the enrichment of predicted target genes of ligands in this set of DE genes compared to the genomic background. Here we choose for a minimum logFC of 0.50, maximum p-value of 0.05, and minimum fraction of expression of 0.05. 

```{r}
logFC_threshold = 0.50
p_val_threshold = 0.05
fraction_cutoff = 0.05
```

We will here choose for applying the p-value cutoff on the p-values corrected for multiple testing because we have many samples in this dataset, so the pseudobulk analysis will less likely be underpowered.

```{r}
p_val_adj = TRUE
empirical_pval = FALSE
```

For the NicheNet ligand-target inference, we also need to select which top n of the predicted target genes will be considered (here: top 250 targets per ligand).

```{r}
top_n_target = 250
```

The NicheNet ligand activity analysis can be run in parallel for each receiver cell type, by changing the number of cores as defined here. This is only recommended if you have many receiver cell type. 

```{r}
cores_system = 8
n.cores = min(cores_system, union(senders_oi, receivers_oi) %>% length()) # use one core per receiver cell type
```

### Define the weights of the prioritization of both expression, differential expression and NicheNet activity information

MultiNicheNet allows the user to define the weights of the following criteria to prioritize ligand-receptor interactions:

* Upregulation of the ligand in a sender cell type and/or upregulation of the receptor in a receiver cell type - in the condition of interest. : `de_ligand` and `de_receptor`
* Sufficiently high expression levels of ligand and receptor in many samples of the same group (to mitigate the influence of outlier samples). : `frac_exprs_ligand_receptor`
* Cell-type and condition specific expression of the ligand in the sender cell type and receptor in the receiver cell type (to mitigate the influence of upregulated but still relatively weakly expressed ligands/receptors) : `exprs_ligand` and `exprs_receptor`
* High NicheNet ligand activity, to further prioritize ligand-receptor pairs based on their predicted effect of the ligand-receptor interaction on the gene expression in the receiver cell type : `activity_scaled` 
* High relative abundance of sender and/or receiver in the condition of interest: `abund_sender` and `abund_receiver` (experimental feature - not recommended to give non-zero weights for default analyses)

The different properties of the sender-ligand---receiver-receptor pairs can be weighted according to the user's preference and insight in the dataset at hand. 

We will set our preference for this dataset as follows - and recommend the user to use the same weights by default if they don't have specific reasons to change some of the parameters:

```{r}
prioritizing_weights_DE = c("de_ligand" = 1,
                         "de_receptor" = 1)
prioritizing_weights_activity = c("activity_scaled" = 2)

prioritizing_weights_expression_specificity = c("exprs_ligand" = 2,
                         "exprs_receptor" = 2)

prioritizing_weights_expression_sufficiency = c("frac_exprs_ligand_receptor" = 1)

prioritizing_weights_relative_abundance = c( "abund_sender" = 0,
                         "abund_receiver" = 0)

```

```{r}
prioritizing_weights = c(prioritizing_weights_DE, 
                         prioritizing_weights_activity, 
                         prioritizing_weights_expression_specificity,
                         prioritizing_weights_expression_sufficiency, 
                         prioritizing_weights_relative_abundance)
```


# Step 2: Perform MultiNicheNet's cell-cell communication analysis

Now we will run the MultiNicheNet wrapper. In the function `multi_nichenet_analysis`, we need to specify that we use one SingleCellExperiment object of which all cell types should be considered as both receiver and sender by setting `sender_receiver_separate = FALSE`. This setting will call the underlying `multi_nichenet_analysis_combined` pipeline function. In case you want to consider different sender and receiver cell types, you should put `sender_receiver_separate = TRUE` and add the additional function arguments as can be seen in `?multi_nichenet_analysis_separate`

To keep track of the different steps, we will here set `verbose = TRUE`

This takes a few minutes for this vignette. The more cell types are considered, the longer this analysis takes.

```{r}
multinichenet_output = multi_nichenet_analysis(sce = sce, celltype_id = celltype_id, sample_id = sample_id, group_id = group_id, 
                                lr_network = lr_network, ligand_target_matrix = ligand_target_matrix, contrasts_oi = contrasts_oi, contrast_tbl = contrast_tbl, batches = batches, covariates = covariates,
                                prioritizing_weights = prioritizing_weights, min_cells = min_cells, logFC_threshold = logFC_threshold, p_val_threshold = p_val_threshold,  
                                fraction_cutoff = fraction_cutoff, p_val_adj = p_val_adj, empirical_pval = empirical_pval, top_n_target = top_n_target, n.cores = n.cores, sender_receiver_separate = FALSE, verbose = TRUE)
```

The output of the MultiNicheNet analysis contains much information. We will now go over this step-by-step

## Check the returned tables in the output

### Average expression value and fraction of each cell type - sample combination

```{r}
multinichenet_output$celltype_info$avg_df %>% head()
multinichenet_output$celltype_info$frq_df %>% head()
multinichenet_output$celltype_info$avg_df_group %>% head()
multinichenet_output$celltype_info$frq_df_group %>% head()
multinichenet_output$celltype_info$rel_abundance_df %>% head()
```

### DE information for each cell type - contrast combination

```{r}
multinichenet_output$celltype_de %>% head()
```

### Output of the NicheNet ligand activity analysis, and the NicheNet ligand-target inference

```{r}
multinichenet_output$ligand_activities_targets_DEgenes$ligand_activities %>% head()
```

### Tables with the final prioritization scores (results per group and per sample)

```{r}
multinichenet_output$prioritization_tables$group_prioritization_tbl %>% head()
```

Based on these prioritization tables, we will define which interactions to visualize in the different plots below.

# Step 3: Visualization of the results of the cell-cell communication analysis

In a first instance, we will look at the broad overview of prioritized interactions via condition-specific Circos plots.

## Circos plot of top-prioritized links

We will look here at the top 50 predictions across all contrasts, senders, and receivers of interest.

```{r}
prioritized_tbl_oi_all = get_top_n_lr_pairs(multinichenet_output$prioritization_tables, 50, rank_per_group = FALSE)
```

```{r, fig.width=15, fig.height=12}
prioritized_tbl_oi = multinichenet_output$prioritization_tables$group_prioritization_tbl %>%
  filter(id %in% prioritized_tbl_oi_all$id) %>%
  distinct(id, sender, receiver, ligand, receptor, group) %>% left_join(prioritized_tbl_oi_all)
prioritized_tbl_oi$prioritization_score[is.na(prioritized_tbl_oi$prioritization_score)] = 0

senders_receivers = union(prioritized_tbl_oi$sender %>% unique(), prioritized_tbl_oi$receiver %>% unique()) %>% sort()

colors_sender = RColorBrewer::brewer.pal(n = length(senders_receivers), name = 'Spectral') %>% magrittr::set_names(senders_receivers)
colors_receiver = RColorBrewer::brewer.pal(n = length(senders_receivers), name = 'Spectral') %>% magrittr::set_names(senders_receivers)

circos_list = make_circos_group_comparison(prioritized_tbl_oi, colors_sender, colors_receiver)
```

## Visualization of scaled ligand-receptor pseudobulk products and ligand activity

Now we will visualize per sample the scaled product of ligand and receptor expression. Samples that were left out of the DE analysis are indicated with a smaller dot (this helps to indicate the samples that did not contribute to the calculation of the logFC, and thus not contributed to the final prioritization)

We will now check the top 50 interactions that were prioritized for idiopathic.pulmonary.fibrosis vs normal:

```{r}
group_oi = "idiopathic.pulmonary.fibrosis"
```

```{r, fig.height=13, fig.width=24}
prioritized_tbl_oi_top_50 = get_top_n_lr_pairs(multinichenet_output$prioritization_tables, 50, groups_oi = group_oi)

plot_oi = make_sample_lr_prod_activity_batch_plots(multinichenet_output$prioritization_tables, prioritized_tbl_oi_top_50, multinichenet_output$grouping_tbl, batches , widths = c(11,1,1))
plot_oi
```

Typically, there are way more than 50 differentially expressed and active ligand-receptor pairs per group across all sender-receiver combinations. Therefore it might be useful to zoom in on specific cell types as senders/receivers:

Eg Fibroblast as receiver:

```{r, fig.height=13, fig.width=24}
prioritized_tbl_oi_top_50 = get_top_n_lr_pairs(multinichenet_output$prioritization_tables, 50, groups_oi = group_oi, receivers_oi = "Fibroblast")

plot_oi = make_sample_lr_prod_activity_batch_plots(multinichenet_output$prioritization_tables, prioritized_tbl_oi_top_50, multinichenet_output$grouping_tbl, batches , widths = c(11,1,1))
plot_oi
```

Eg Fibroblast as sender:

```{r, fig.height=13, fig.width=24}
prioritized_tbl_oi_top_50 = get_top_n_lr_pairs(multinichenet_output$prioritization_tables, 50, groups_oi = group_oi, senders_oi = "Fibroblast")

plot_oi = make_sample_lr_prod_activity_batch_plots(multinichenet_output$prioritization_tables, prioritized_tbl_oi_top_50, multinichenet_output$grouping_tbl, batches , widths = c(11,1,1))
plot_oi
```

You can make these plots also for the other groups, like we will illustrate now for the normal group

```{r}
group_oi = "normal"
```

```{r, fig.height=13, fig.width=24}
prioritized_tbl_oi_top_50 = get_top_n_lr_pairs(multinichenet_output$prioritization_tables, 50, groups_oi = group_oi)

plot_oi = make_sample_lr_prod_activity_batch_plots(multinichenet_output$prioritization_tables, prioritized_tbl_oi_top_50, multinichenet_output$grouping_tbl, batches , widths = c(11,1,1))
plot_oi
```


## Intercellular regulatory network systems view

As additional plot, we can generate a 'systems' view of these intercellular feedback and cascade processes than can be occuring between the different cell populations involved. In this plot, we will draw links between ligands of sender cell types their ligand/receptor-annotated target genes in receiver cell types. So links are ligand-target links (= gene regulatory links) and not ligand-receptor protein-protein interactions!

```{r}
prioritized_tbl_oi = get_top_n_lr_pairs(multinichenet_output$prioritization_tables, 150, rank_per_group = FALSE)

lr_target_prior_cor_filtered = multinichenet_output$prioritization_tables$group_prioritization_tbl$group %>% unique() %>% lapply(function(group_oi){
  lr_target_prior_cor_filtered = multinichenet_output$lr_target_prior_cor %>% inner_join(multinichenet_output$ligand_activities_targets_DEgenes$ligand_activities %>% distinct(ligand, target, direction_regulation, contrast)) %>% inner_join(contrast_tbl) %>% filter(group == group_oi)
  lr_target_prior_cor_filtered_up = lr_target_prior_cor_filtered %>% filter(direction_regulation == "up") %>% filter( (rank_of_target < top_n_target) & (pearson > 0.50 | spearman > 0.50))
  lr_target_prior_cor_filtered_down = lr_target_prior_cor_filtered %>% filter(direction_regulation == "down") %>% filter( (rank_of_target < top_n_target) & (pearson < -0.50 | spearman < -0.50))
  lr_target_prior_cor_filtered = bind_rows(lr_target_prior_cor_filtered_up, lr_target_prior_cor_filtered_down)
}) %>% bind_rows()

```

```{r, fig.width=30, fig.height=12}
graph_plot = make_ggraph_ligand_target_links(lr_target_prior_cor_filtered = lr_target_prior_cor_filtered, prioritized_tbl_oi = prioritized_tbl_oi, colors = colors_sender)
graph_plot$plot
graph_plot$source_df_lt %>% head()
graph_plot$nodes_df %>% head()
```

## Comparing results of the analysis with batch correction to an analysis where we would ignore the existence of the batches.

To assess how much difference the batch correction made, we will now run the MultiNicheNet analysis without batch correction
```{r}
batches =  NA

multinichenet_output_noBC = multi_nichenet_analysis(sce = sce, celltype_id = celltype_id, sample_id = sample_id, group_id = group_id, 
                                lr_network = lr_network, ligand_target_matrix = ligand_target_matrix, contrasts_oi = contrasts_oi, contrast_tbl = contrast_tbl, batches = batches, covariates = covariates,
                                prioritizing_weights = prioritizing_weights, min_cells = min_cells, logFC_threshold = logFC_threshold, p_val_threshold = p_val_threshold,  
                                fraction_cutoff = fraction_cutoff, p_val_adj = p_val_adj, empirical_pval = empirical_pval, top_n_target = top_n_target, n.cores = n.cores, sender_receiver_separate = FALSE, verbose = TRUE)
```
First we will show the top 50 interactions from the batch-corrected analysis (BC analysis) with non-corrected pseudobulk expression values

```{r}
group_oi = "idiopathic.pulmonary.fibrosis"
batches =  "dataset_origin"
```

```{r, fig.height=13, fig.width=24}
prioritized_tbl_oi_top_50 = get_top_n_lr_pairs(multinichenet_output$prioritization_tables, 50, groups_oi = group_oi) # from BC-analysis

plot_oi = make_sample_lr_prod_activity_batch_plots(multinichenet_output_noBC$prioritization_tables, prioritized_tbl_oi_top_50, multinichenet_output$grouping_tbl, batches , widths = c(11,1,1))
plot_oi
```

For this dataset and these cell types, the non-corrected expression values look similar to the corrected ones.

Now we will combine the prioritization tables of both analyses. This will then later be used to define which interactions are most specific to the BC analysis compared to the non-BC analysis

```{r}
prioritized_tbl_oi_high5000_withBC = get_top_n_lr_pairs(multinichenet_output$prioritization_tables, 5000, rank_per_group = FALSE)
prioritized_tbl_oi_high5000_noBC = get_top_n_lr_pairs(multinichenet_output_noBC$prioritization_tables, 5000, rank_per_group = FALSE)

prioritized_tbl_oi_high5000 = prioritized_tbl_oi_high5000_withBC %>% rename(score_BC = prioritization_score , rank_BC = prioritization_rank) %>% full_join(prioritized_tbl_oi_high5000_noBC %>% rename(score_noBC = prioritization_score , rank_noBC = prioritization_rank))

head(prioritized_tbl_oi_high5000) # you can see some interactions NA in an analysis: this because ligand or receptor gene was filtered out

prioritized_tbl_oi_high5000 = prioritized_tbl_oi_high5000 %>% mutate(diff_rank = rank_BC  - rank_noBC, diff_score = score_BC - score_noBC) 
```

Inspecting `prioritized_tbl_oi_high5000` enables you to see how prioritization scores and ranks differ between both analyses.

```{r}
prioritized_tbl_oi_high5000 %>% View()
```

Now we will define which interactions are most specific to the BC analysis compared to the non-BC analysis

```{r}
BC_interactions = prioritized_tbl_oi_high5000 %>% arrange(-diff_score) %>% filter(group == group_oi) %>% filter(rank_BC < 2000) %>% pull(id) %>% head(10)
noBC_interactions = prioritized_tbl_oi_high5000 %>% arrange(diff_score) %>% filter(group == group_oi) %>% filter(rank_noBC < 2000) %>% pull(id) %>% head(10)

BC_interactions_df = prioritized_tbl_oi_high5000 %>% arrange(-score_BC) %>% filter(is.na(diff_rank)) %>% filter(group == group_oi) %>% head(10)
noBC_interactions_df = prioritized_tbl_oi_high5000 %>% arrange(-score_noBC)  %>% filter(is.na(diff_rank)) %>% filter(group == group_oi) %>% head(10)

BC_specific_interactions_df = prioritized_tbl_oi_high5000 %>% arrange(-score_BC) %>% filter(group == group_oi) %>% filter(id %in% BC_interactions)
noBC_specific_interactions_df = prioritized_tbl_oi_high5000 %>% arrange(-score_noBC) %>% filter(group == group_oi) %>% filter(id %in% noBC_interactions)
```

### Visualization of scaled_LR_prod_activity per sample

#### BC-specific hits with BC expression values plot

```{r, fig.height=9, fig.width=27}
plot_oi = make_sample_lr_prod_activity_batch_plots(multinichenet_output$prioritization_tables, BC_specific_interactions_df, multinichenet_output$grouping_tbl, batches , widths = c(11,1,1))
plot_oi
```

#### BC-specific hits with noBC expression values plot

```{r, fig.height=9, fig.width=27}
plot_oi = make_sample_lr_prod_activity_batch_plots(multinichenet_output_noBC$prioritization_tables,  BC_specific_interactions_df, multinichenet_output$grouping_tbl, batches, widths = c(11,1,1))
plot_oi
```


#### noBC-specific hits with BC expression values plot

```{r, fig.height=9, fig.width=27}
plot_oi = make_sample_lr_prod_activity_batch_plots(multinichenet_output_noBC$prioritization_tables, noBC_specific_interactions_df, multinichenet_output$grouping_tbl, batches , widths = c(11,1,1))
plot_oi
```

#### noBC-specific hits with no BC expression values plots

```{r, fig.height=9, fig.width=27}
plot_oi = make_sample_lr_prod_activity_batch_plots(multinichenet_output_noBC$prioritization_tables, noBC_specific_interactions_df , multinichenet_output$grouping_tbl, batches , widths = c(11,1,1))
plot_oi
```

Important to notice: specific interactions of batch or noBATCH analysis: also including the ligand activity metric, so not only the DE of the LR pair!

### Check target genes!

Show now also DE genes that are different! -  this is important for the ligand activities!


```{r, fig.height=9, fig.width=27}
pval_df_targets = multinichenet_output_noBC$celltype_de %>% filter(logFC > 0 & contrast == "idiopathic.pulmonary.fibrosis-normal") %>% select(gene, cluster_id, p_adj) %>% distinct() %>% rename(p_adj_noBC = p_adj) %>% 
  inner_join(multinichenet_output$celltype_de %>% filter(logFC > 0 & contrast == "idiopathic.pulmonary.fibrosis-normal") %>% select(gene, cluster_id, p_adj) %>% distinct() %>% rename(p_adj_BC = p_adj))

## targets with more DE in noBC condition
pval_df_targets %>% mutate(prop = p_adj_BC/p_adj_noBC)  %>% arrange(-prop) %>% filter(prop > 1 & p_adj_noBC < 0.05 & p_adj_BC > 0.05) 
pval_df_targets %>% mutate(prop = p_adj_BC/p_adj_noBC)  %>% arrange(-prop) %>% filter(prop > 1 & p_adj_noBC < 0.05 & p_adj_BC > 0.05) %>% group_by(cluster_id) %>% count() %>% arrange(-n)

## targets with more DE in BC condition
pval_df_targets %>% mutate(prop = p_adj_BC/p_adj_noBC)  %>% arrange(-prop) %>% filter(prop < 1 & p_adj_noBC > 0.05 & p_adj_BC < 0.05) 
pval_df_targets %>% mutate(prop = p_adj_BC/p_adj_noBC)  %>% arrange(-prop) %>% filter(prop < 1 & p_adj_noBC > 0.05 & p_adj_BC < 0.05) %>% group_by(cluster_id) %>% count() %>% arrange(-n)
```

Define BC-specific DE genes in Fibroblasts

```{r, fig.height=11, fig.width=17}
targets_oi = pval_df_targets %>% mutate(prop = p_adj_BC/p_adj_noBC)  %>% arrange(-prop) %>% filter(prop < 1 & p_adj_noBC > 0.05 & p_adj_BC < 0.05) %>% filter(cluster_id == "Fibroblast") %>% pull(gene)
targets_oi %>% tibble(gene = .) 
```



Visualize these BC_specific_targets with corrected expression values:

```{r, fig.height=11, fig.width=17}
p_target = make_DEgene_dotplot_pseudobulk_batch(genes_oi = targets_oi, celltype_info = multinichenet_output$celltype_info, prioritization_tables = multinichenet_output$prioritization_tables, celltype_oi = "Fibroblast", batch_oi = batches, grouping_tbl = multinichenet_output$grouping_tbl)
p_target$pseudobulk_plot
```

Visualize these BC_specific_targets with non-corrected expression values:

```{r, fig.height=11, fig.width=17}
p_target = make_DEgene_dotplot_pseudobulk_batch(genes_oi = targets_oi, celltype_info = multinichenet_output_noBC$celltype_info, prioritization_tables = multinichenet_output_noBC$prioritization_tables, celltype_oi = "Fibroblast",  batch_oi = batches, grouping_tbl = multinichenet_output$grouping_tbl)
p_target$pseudobulk_plot
```

Here you see a clear difference, pointing to the need to properly correct for batch effects to get a good set of DE genes, which will be used in the ligand activity calculations within MultiNicheNet.