---
title: 'Multi-Group Multi-Sample Cell-Cell Communication Analysis via NicheNet: CRC
  application'
author: "Robin Browaeys"
date: "5-2-2021"
output: html_document
---


```{r setup, include = FALSE}
# knitr::opts_knit$set(
#   collapse = TRUE,
#   # comment = "#>",
#   warning = FALSE,
#   message = FALSE,
#   # root.dir = 'C:/Users/rbrowaey/work/Research/NicheNet/current_projects/CRC_NicheNet'
#     root.dir = '../'
# 
# )
knitr::opts_knit$set(root.dir = 'C:/Users/rbrowaey/work/Research/NicheNet/current_projects/CRC_NicheNet' )

```

In this analysis we will analyze differential cell-cell communication between different patient groups in colorectal cancer.

# Step 0: Load packages and NicheNet ligand-receptor network and ligand-target matrix

```{r}
library(Seurat)
library(BiocStyle)
library(Nebulosa)
library(cowplot)
library(tidyverse)
library(limma)
library(muscat)
library(ExperimentHub)
library(scater)
library(circlize)
library(RColorBrewer)
source("scripts/functions_muscat_de.R")
```

```{r}
# LR network
lr_network = readRDS(url("https://zenodo.org/record/3260758/files/lr_network.rds"))
lr_network = lr_network %>% filter(! database %in% c("ppi_prediction","ppi_prediction_go"))
lr_network = lr_network %>% dplyr::rename(ligand = from, receptor = to) %>% distinct(ligand, receptor)
```

```{r}
ligand_target_matrix = readRDS(url("https://zenodo.org/record/3260758/files/ligand_target_matrix.rds"))
```

# Step 1: Prepare Seurat Objects for Sender and Receiver cells

## T cells as receiver

Set the cell type label of interest as cell identity and keep only good samples.
In this analysis we are only interested in MSI, CMS2 and CMS4 patients, therefore we will pool all the other samples so that we don't remove cell types that have too few cells in patient groups that are not of interest to our comparison.

```{r}
getwd()
seurat_obj_receiver = readRDS("data/201013KULSMCAllTFiltered.rds")
seurat_obj_receiver@meta.data$TmajorScaled = seurat_obj_receiver@meta.data$TmajorScaled %>% make.names()
seurat_obj_receiver = SetIdent(seurat_obj_receiver, value = "TmajorScaled")
good_samples = c("EXT001","EXT002","EXT003","EXT009","EXT010","EXT011","EXT012","EXT013","EXT031","EXT032","EXT048","EXT049","EXT050","EXT093","EXT094","EXT095","EXT097","EXT098","EXT099","EXT121","EXT122","EXT123","EXT127","EXT128","EXT129","PM0001N","PM0001T","PM0002N","PM0002T","PM0003T","PM0004N", "PM0004T","PM0005N","PM0005T","PM0010N","PM0010T","PM0011T","PM0014T","PM0016T","PM0017T","PM0018T","PM0019T","PM0020T","PM0021T","PM0022T","PM0023T","PM0024T","PM0025T","SMC0010T","SMC057T","SMC063T","SMC071T","SMC083T","SMC105T","SMC108T","SMC112T","SMC119T","SMC125T","SMC126T","SMC136T","SMC145T","SMC164T","SMC174T","SMC179T","SMC180T","SMC197T","SMC200T","SMC215T","SMC216T","SMC217T")
seurat_obj_receiver = subset(seurat_obj_receiver, subset = sample_ID %in% good_samples)
seurat_obj_receiver@meta.data$CMSMSItis[seurat_obj_receiver@meta.data$CMSMSItis == "NA"] = "CMS_NA" # because NA as name will result in errors later on

# because our contrasts of interest have to do with CMS2-CMS4-MSI, we will pool the groups that are not of interest! 
# This will help us in having more cell types for which we can do the DE analysis
seurat_obj_receiver@meta.data$CMSMSItis[seurat_obj_receiver@meta.data$CMSMSItis == "CMS_NA"] = "Other" # because NA as name will result in errors later on
seurat_obj_receiver@meta.data$CMSMSItis[seurat_obj_receiver@meta.data$CMSMSItis == "CMS1"] = "Other" # because NA as name will result in errors later on
seurat_obj_receiver@meta.data$CMSMSItis[seurat_obj_receiver@meta.data$CMSMSItis == "CMS3"] = "Other" # because NA as name will result in errors later on
seurat_obj_receiver@meta.data$CMSMSItis[seurat_obj_receiver@meta.data$CMSMSItis == "Normal"] = "Other" # because NA as name will result in errors later on
table(seurat_obj_receiver@meta.data$CMSMSItis, seurat_obj_receiver@meta.data$TmajorScaled)
DimPlot(seurat_obj_receiver)
```

## Myeloid cells as receiver
```{r}
getwd()
seurat_obj_sender = readRDS("data/201111KULSMCallMyeloid.rds")
seurat_obj_sender@meta.data$FinalMyeloidLabel0.4 = seurat_obj_sender@meta.data$FinalMyeloidLabel0.4 %>% make.names()
seurat_obj_sender@meta.data$sample_ID = seurat_obj_sender@meta.data$sample_ID %>% toupper()
seurat_obj_sender = SetIdent(seurat_obj_sender, value = "FinalMyeloidLabel0.4")
good_samples = c("EXT001","EXT002","EXT003","EXT009","EXT010","EXT011","EXT012","EXT013","EXT031","EXT032","EXT048","EXT049","EXT050","EXT093","EXT094","EXT095","EXT097","EXT098","EXT099","EXT121","EXT122","EXT123","EXT127","EXT128","EXT129","PM0001N","PM0001T","PM0002N","PM0002T","PM0003T","PM0004N", "PM0004T","PM0005N","PM0005T","PM0010N","PM0010T","PM0011T","PM0014T","PM0016T","PM0017T","PM0018T","PM0019T","PM0020T","PM0021T","PM0022T","PM0023T","PM0024T","PM0025T","SMC0010T","SMC057T","SMC063T","SMC071T","SMC083T","SMC105T","SMC108T","SMC112T","SMC119T","SMC125T","SMC126T","SMC136T","SMC145T","SMC164T","SMC174T","SMC179T","SMC180T","SMC197T","SMC200T","SMC215T","SMC216T","SMC217T")
seurat_obj_sender = subset(seurat_obj_sender, subset = sample_ID %in% good_samples)
seurat_obj_sender@meta.data$msitis %>% table()
seurat_obj_sender@meta.data$CMSMSItis[seurat_obj_sender@meta.data$CMSMSItis == "MSS"] = "CMS_NA" # because NA as name will result in errors later on
# seurat_obj_sender@meta.data$FinalMyeloidLabel0.4[seurat_obj_sender@meta.data$FinalMyeloidLabel0.4 == "Pro.inflammatory.myeloid"] = "Monocytes"
# seurat_obj_sender@meta.data$FinalMyeloidLabel0.4[seurat_obj_sender@meta.data$FinalMyeloidLabel0.4 == "Migratory.DC"] = "cDC2"
seurat_obj_sender@meta.data$CMSMSItis[seurat_obj_sender@meta.data$CMSMSItis == "CMS_NA"] = "Other" # because NA as name will result in errors later on
seurat_obj_sender@meta.data$CMSMSItis[seurat_obj_sender@meta.data$CMSMSItis == "CMS1"] = "Other" # because NA as name will result in errors later on
seurat_obj_sender@meta.data$CMSMSItis[seurat_obj_sender@meta.data$CMSMSItis == "CMS3"] = "Other" # because NA as name will result in errors later on
seurat_obj_sender@meta.data$CMSMSItis[seurat_obj_sender@meta.data$CMSMSItis == "Normal"] = "Other" # because NA as name will result in errors later on
table(seurat_obj_sender@meta.data$CMSMSItis, seurat_obj_sender@meta.data$FinalMyeloidLabel0.4)
DimPlot(seurat_obj_sender)
```

# Step 2: Prepare the cell-cell communication analysis

## Define senders and receivers of interest

```{r}
senders_oi = c("Monocytes","cDC1","cDC2","Macrophages","CD3.myeloid.diverse","proliferating","Pro.inflammatory.myeloid","Migratory.DC","pDC")
receivers_oi = c("CD4","CD8")

seurat_obj_sender = seurat_obj_sender %>% subset(idents = senders_oi)
seurat_obj_receiver = seurat_obj_receiver %>% subset(idents = receivers_oi)
```

## Define in which metadata columns we can find the group, sample and cell type IDs

```{r}
sample_id = "sample_ID"
group_id = "CMSMSItis"
celltype_id_receiver = "TmajorScaled"
celltype_id_sender = "FinalMyeloidLabel0.4"
```

## Define the contrasts and covariates of interest for the DE analysis, and the minimal number of cells of a cell type that each sample should have.

```{r}
covariates = c("dataset")
contrasts_oi = c("'CMS4-(CMS2+MSI)/2','CMS2-(CMS4+MSI)/2','MSI-(CMS2+CMS4)/2'")
contrast_tbl = tibble(contrast = 
                        c("CMS4-(CMS2+MSI)/2","CMS2-(CMS4+MSI)/2","MSI-(CMS2+CMS4)/2"), 
                      group = c("CMS4","CMS2","MSI"))
min_cells = 10

```

## Define the parameters for the NicheNet ligand activity analysis 

```{r}
logFC_threshold = 0.33
p_val_threshold = 0.05
frac_cutoff = 0.05
p_val_adj = FALSE 
top_n_target = 500
```


## Define the weights of the prioritization of both expression, differential expression and NicheNet activity information


```{r}
prioritizing_weights = c("scaled_lfc_ligand" = 1,
                         "scaled_p_val_ligand" = 1,
                         "scaled_lfc_receptor" = 1,
                         "scaled_p_val_receptor" = 1,
                         "scaled_activity_scaled" = 1.5,
                         "scaled_activity" = 0.5,
                         "scaled_avg_exprs_ligand" = 1,
                         "scaled_avg_frq_ligand" = 1,
                         "scaled_avg_exprs_receptor" = 1,
                         "scaled_avg_frq_receptor" = 1,
                         "fraction_expressing_ligand_receptor" = 1,
                         "scaled_abundance_sender" = 0,
                         "scaled_abundance_receiver" = 0)
```

# Step 3: Perform the cell-cell communication analysis

```{r}
output = multi_nichenet_analysis(seurat_obj_receiver = seurat_obj_receiver, seurat_obj_sender = seurat_obj_sender, 
                                celltype_id_receiver = celltype_id_receiver, celltype_id_sender = celltype_id_sender,sample_id = sample_id, group_id = group_id, 
                                lr_network = lr_network, ligand_target_matrix = ligand_target_matrix, contrasts_oi = contrasts_oi, contrast_tbl = contrast_tbl, 
                                prioritizing_weights = prioritizing_weights, min_cells = min_cells, logFC_threshold = logFC_threshold, p_val_threshold = p_val_threshold,  
                                frac_cutoff = frac_cutoff, p_val_adj = p_val_adj, top_n_target = top_n_target)

```

```{r}
saveRDS(output, "output/multi_nichenet_output")
```

# Step 4: Visualize the results of the cell-cell communication analysis

## Visualization of scaled_LR_prod (average expression) and scaled_LR_frac (average fraction) per sample

```{r}
group_oi = "MSI"
```

```{r, fig.height=18, fig.width=12}
prioritized_tbl_oi = output$prioritization_tables$group_prioritization_tbl %>% 
  distinct(id, sender, receiver, lr_interaction, group, ligand_receptor_lfc_avg, activity_scaled, fraction_expressing_ligand_receptor,  prioritization_score) %>% 
  filter(ligand_receptor_lfc_avg > 0 & fraction_expressing_ligand_receptor > 0) %>% 
  filter(group == group_oi) %>% group_by(group) %>% top_n(75, prioritization_score) 

plot_oi = make_sample_lr_prod_plots(output$prioritization_tables, prioritized_tbl_oi)
plot_oi
```

## Visualization of expression-logFC per group and ligand activity

```{r}
receiver_oi = "CD4"
group_oi = "MSI"
```

```{r, fig.width=15, fig.height=12}

prioritized_tbl_oi = output$prioritization_tables$group_prioritization_tbl %>% 
  filter(fraction_ligand_group > frac_cutoff & fraction_receptor_group > frac_cutoff) %>% 
  distinct(id, sender, receiver, lr_interaction, group, ligand_receptor_lfc_avg, activity_scaled, fraction_ligand_group, fraction_expressing_ligand_receptor, scaled_avg_exprs_ligand, prioritization_score) %>% 
  filter(ligand_receptor_lfc_avg > 0 & fraction_expressing_ligand_receptor > 0) %>% 
  filter(group == group_oi & receiver == receiver_oi) %>% top_n(100, prioritization_score) 

plot_oi = make_group_lfc_exprs_activity_plot(output$prioritization_tables, prioritized_tbl_oi, receiver_oi)
plot_oi

```

## Circos plot of top-prioritized links

```{r, fig.width=15, fig.height=12}
prioritized_tbl_oi = output$prioritization_tables$group_prioritization_tbl %>% 
  filter(fraction_ligand_group > frac_cutoff & fraction_receptor_group > frac_cutoff) %>% 
  distinct(id, sender, receiver, ligand, receptor, group, prioritization_score, ligand_receptor_lfc_avg, fraction_expressing_ligand_receptor) %>% 
  filter(ligand_receptor_lfc_avg > 0 & fraction_expressing_ligand_receptor > 0) %>%   group_by(group) %>%  top_n(25, prioritization_score) 

prioritized_tbl_oi = output$prioritization_tables$group_prioritization_tbl %>% 
  filter(id %in% prioritized_tbl_oi$id) %>% 
  distinct(id, sender, receiver, ligand, receptor, group) %>% left_join(prioritized_tbl_oi)
prioritized_tbl_oi$prioritization_score[is.na(prioritized_tbl_oi$prioritization_score)] = 0


n_senders = prioritized_tbl_oi$sender %>% unique() %>% length()
n_receivers = prioritized_tbl_oi$receiver %>% unique() %>% length()

colors_sender = c("orange","tomato","red","violetred3", "orchid2","mediumpurple1","steelblue2", "royalblue")
# colors_sender = brewer.pal(8, "Spectral")
colors_receiver = c("limegreen","darkolivegreen2")

circos_list = make_circos_group_comparison(prioritized_tbl_oi, colors_sender, colors_receiver)

```

## Single-cell-based Nebulosa, Feature, and Violin plots of ligand-receptor interaction of interest

```{r}
prioritized_tbl_oi = output$prioritization_tables$group_prioritization_tbl %>% 
  filter(fraction_ligand_group > frac_cutoff & fraction_receptor_group > frac_cutoff) %>% 
  distinct(id, sender, receiver, ligand, receptor, group, prioritization_score) %>% 
  group_by(group, receiver) %>% top_n(5, prioritization_score) 
prioritized_tbl_oi
```

```{r}
ligand_oi = "TNFSF15"
receptor_oi = "TNFRSF25"
group_oi = "MSI"
sender_oi = "Pro.inflammatory.myeloid"
receiver_oi = "CD4"
```

```{r, fig.width=9, fig.height=7}
plot_list = make_ligand_receptor_nebulosa_feature_plot(seurat_obj_sender, seurat_obj_receiver, ligand_oi, receptor_oi, group_oi, group_id, celltype_id_sender, celltype_id_receiver, senders_oi, receivers_oi, prioritized_tbl_oi)
plot_list$nebulosa
plot_list$feature
```

```{r, fig.width=9, fig.height=7}
plot_list2 = make_ligand_receptor_violin_plot(seurat_obj_sender, seurat_obj_receiver, ligand_oi, receptor_oi, sender_oi, receiver_oi, group_oi, group_id, sample_id, celltype_id_sender, celltype_id_receiver, prioritized_tbl_oi)
plot_list2$violin_group
plot_list2$violin_sample
```

## Visualization of ligand-activity, ligand-target links, and target gene expression

```{r}
group_oi = "MSI"
receiver_oi = "CD4"
prioritized_tbl_oi = output$prioritization_tables$group_prioritization_tbl %>% 
  filter(fraction_ligand_group > frac_cutoff & fraction_receptor_group > frac_cutoff) %>% 
  distinct(id, sender, receiver, ligand, receptor, group, prioritization_score, ligand_receptor_lfc_avg, fraction_expressing_ligand_receptor, activity_scaled) %>% 
  filter(ligand_receptor_lfc_avg > 0 & fraction_expressing_ligand_receptor > 0) %>% 
  filter(group == group_oi & receiver == receiver_oi) %>% 
  group_by(group) %>% top_n(100, prioritization_score) %>% top_n(50, activity_scaled) %>% arrange(-activity_scaled)
```

  
```{r, fig.width=16, fig.height=13}
combined_plot = make_ligand_activity_target_plot(group_oi, receiver_oi, prioritized_tbl_oi, output$ligand_activities_targets_DEgenes, contrast_tbl, output$grouping_tbl, output$receiver_info)
combined_plot
```

## Make target gene violin and nebulosa plots

```{r}
receiver_oi = "CD4"
group_oi = "MSI"

output$ligand_activities_targets_DEgenes$de_genes_df %>% inner_join(contrast_tbl) %>% filter(group == group_oi) %>% arrange(p_val) %>% filter(receiver == receiver_oi) %>% top_n(100, logFC) %>% top_n(5, max_frac) 
```

  
```{r, fig.width=14, fig.height=8}
target_oi = "CST7"

make_target_violin_plot(seurat_obj_receiver, target_oi, receiver_oi, group_oi, group_id, sample_id, celltype_id_receiver, output$prioritization_tables$group_prioritization_tbl)
make_target_nebulosa_feature_plot(seurat_obj_receiver, target_oi, group_oi, group_id, celltype_id_receiver, receivers_oi, output$prioritization_tables$group_prioritization_tbl) 

```

## Make Dotplot for all DE genes/targets

```{r}
receiver_oi = "CD4"
group_oi = "MSI"

targets_oi1 = output$ligand_activities_targets_DEgenes$de_genes_df %>% inner_join(contrast_tbl) %>% filter(group == group_oi) %>% arrange(p_val) %>% filter(receiver == receiver_oi) %>% pull(gene) %>% unique()

targets_oi2 = output$ligand_activities_targets_DEgenes$de_genes_df %>% inner_join(contrast_tbl) %>% filter(group == group_oi) %>% arrange(p_val) %>% filter(receiver == receiver_oi) %>% top_n(100, -p_val) %>% top_n(50, logFC)  %>% pull(gene) 

targets_oi = intersect(targets_oi1, targets_oi2)

```

```{r, fig.width=12, fig.height=15}
p_target = make_sample_target_plots(output$receiver_info, targets_oi, receiver_oi, output$grouping_tbl %>% filter(group %in% c("CMS2","CMS4","MSI")))
p_target
```

Example figure to demonstrate how to add covariates

```{r}
p_covariate = output$grouping_tbl %>% filter(group %in% c("CMS2","CMS4","MSI")) %>% mutate(dataset_ = " Dataset ") %>% 
    ggplot(aes(sample, dataset_, fill = dataset)) +
    geom_tile(color = "black") +
    facet_grid(.~group, scales = "free", space = "free") +
    scale_x_discrete(position = "top") +
    theme_light() +
    theme(
      axis.ticks = element_blank(),
      axis.title.x = element_text(size = 0),
      axis.title.y = element_text(size = 0),
      axis.text.y = element_text(face = "bold", size = 9),
      axis.text.x = element_text(size = 9,  angle = 90,hjust = 0),
      strip.text.x.top = element_text(angle = 0),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.spacing.x = unit(2.5, "lines"),
      panel.spacing.y = unit(0.25, "lines"),
      strip.text.x = element_text(size = 11, color = "black", face = "bold"),
      strip.text.y = element_text(size = 9, color = "black", face = "bold", angle = 0),
      strip.background = element_rect(color="darkgrey", fill="whitesmoke", size=1.5, linetype="solid")
    ) + scale_fill_brewer(palette = "Set2")
p_covariate
```

```{r, fig.width=12, fig.height=16}
plot = patchwork::wrap_plots(p_covariate, p_target, nrow = 2, heights = c(2,length(targets_oi)), guides = "collect")
plot
```

 