---
title: 'Multi-Group Multi-Sample Cell-Cell Communication analysis via NicheNet: HNSCC
  application - all-vs-all'
author: "Robin Browaeys"
date: "9-2-2021"
output: html_document
---

```{r setup, include = FALSE}
# knitr::opts_knit$set(
#   collapse = TRUE,
#   # comment = "#>",
#   warning = FALSE,
#   message = FALSE,
#   # root.dir = 'C:/Users/rbrowaey/work/Research/NicheNet/current_projects/CRC_NicheNet'
#     root.dir = '../'
# 
# )
knitr::opts_knit$set(root.dir = 'C:/Users/rbrowaey/work/Research/NicheNet/current_projects/CRC_NicheNet' )

```

In this analysis we will analyze differential cell-cell communication between different patient groups in colorectal cancer. In contrast to the other vignette, this vignette will show how to do an all-vs-all analysis in which you consider all cell types of interest as possible senders and receivers.

# Step 0: Load packages and NicheNet ligand-receptor network and ligand-target matrix

```{r}
library(Seurat)
library(BiocStyle)
library(Nebulosa)
library(cowplot)
library(tidyverse)
library(limma)
library(muscat)
library(ExperimentHub)
library(scater)
library(circlize)
library(RColorBrewer)
source("scripts/functions_muscat_de.R")
```

```{r}
# LR network
lr_network = readRDS(url("https://zenodo.org/record/3260758/files/lr_network.rds"))
lr_network = lr_network %>% filter(! database %in% c("ppi_prediction","ppi_prediction_go"))
lr_network = lr_network %>% dplyr::rename(ligand = from, receptor = to) %>% distinct(ligand, receptor)
```

```{r}
ligand_target_matrix = readRDS(url("https://zenodo.org/record/3260758/files/ligand_target_matrix.rds"))
```

# Step 1: Prepare Seurat Objects for Sender and Receiver cells


```{r}
getwd()
seurat_obj = readRDS("data/seurat_obj_hnscc.rds") 
DimPlot(seurat_obj)
DimPlot(seurat_obj, group.by = "pEMT")
DimPlot(seurat_obj, group.by = "pEMT_fine")
```

# Step 2: Prepare the cell-cell communication analysis


## Define in which metadata columns we can find the group, sample and cell type IDs

```{r}
sample_id = "tumor"
group_id = "pEMT"
celltype_id = "celltype"
```

## Define the contrasts and covariates of interest for the DE analysis, and the minimal number of cells of a cell type that each sample should have.

```{r}
covariates = NA
contrasts_oi = c("'High-Low','Low-High'")
contrast_tbl = tibble(contrast = 
                        c("High-Low","Low-High"), 
                      group = c("High","Low"))
min_cells = 1

```

## Define the parameters for the NicheNet ligand activity analysis 

```{r}
logFC_threshold = 1
p_val_threshold = 0.05
fraction_cutoff = 0.05
p_val_adj = FALSE 
top_n_target = 250
```


## Define the weights of the prioritization of both expression, differential expression and NicheNet activity information


```{r}
prioritizing_weights = c("scaled_lfc_ligand" = 1,
                         "scaled_p_val_ligand" = 1,
                         "scaled_lfc_receptor" = 1,
                         "scaled_p_val_receptor" = 1,
                         "scaled_activity_scaled" = 1.5,
                         "scaled_activity" = 0.5,
                         "scaled_avg_exprs_ligand" = 1,
                         "scaled_avg_frq_ligand" = 1,
                         "scaled_avg_exprs_receptor" = 1,
                         "scaled_avg_frq_receptor" = 1,
                         "fraction_expressing_ligand_receptor" = 1,
                         "scaled_abundance_sender" = 0,
                         "scaled_abundance_receiver" = 0)
```

# Step 3: Perform the cell-cell communication analysis

sender_receiver_separate = FALSE: to demonstrate that our senders and receivers are in the same object

```{r}
output = multi_nichenet_analysis(seurat_obj = seurat_obj, celltype_id = celltype_id, sample_id = sample_id, group_id = group_id, 
                                lr_network = lr_network, ligand_target_matrix = ligand_target_matrix, contrasts_oi = contrasts_oi, contrast_tbl = contrast_tbl, 
                                prioritizing_weights = prioritizing_weights, min_cells = min_cells, logFC_threshold = logFC_threshold, p_val_threshold = p_val_threshold,  
                                fraction_cutoff = fraction_cutoff, p_val_adj = p_val_adj, top_n_target = top_n_target, sender_receiver_separate = FALSE)

```

# Step 4: Visualize the results of the cell-cell communication analysis

## Visualization of scaled_LR_prod (average expression) and scaled_LR_frac (average fraction) per sample

```{r}
group_oi = "High"
```

```{r, fig.height=18, fig.width=12}
prioritized_tbl_oi = output$prioritization_tables$group_prioritization_tbl %>% 
  distinct(id, sender, receiver, lr_interaction, group, ligand_receptor_lfc_avg, activity_scaled, fraction_expressing_ligand_receptor,  prioritization_score) %>% 
  filter(ligand_receptor_lfc_avg > 0 & fraction_expressing_ligand_receptor > 0) %>% 
  filter(group == group_oi) %>% group_by(group) %>% top_n(75, prioritization_score) 

plot_oi = make_sample_lr_prod_plots(output$prioritization_tables, prioritized_tbl_oi)
plot_oi
```

## Visualization of expression-logFC per group and ligand activity

```{r}
receiver_oi = "Malignant"
group_oi = "High"
```

```{r, fig.width=15, fig.height=12}

prioritized_tbl_oi = output$prioritization_tables$group_prioritization_tbl %>% 
  filter(fraction_ligand_group > fraction_cutoff & fraction_receptor_group > fraction_cutoff) %>% 
  distinct(id, sender, receiver, lr_interaction, group, ligand_receptor_lfc_avg, activity_scaled, fraction_ligand_group, fraction_expressing_ligand_receptor, scaled_avg_exprs_ligand, prioritization_score) %>% 
  filter(ligand_receptor_lfc_avg > 0 & fraction_expressing_ligand_receptor > 0) %>% 
  filter(group == group_oi & receiver == receiver_oi) %>% top_n(75, prioritization_score) 

plot_oi = make_group_lfc_exprs_activity_plot(output$prioritization_tables, prioritized_tbl_oi, receiver_oi)
plot_oi

```


## Circos plot of top-prioritized links

```{r, fig.width=15, fig.height=12}
prioritized_tbl_oi = output$prioritization_tables$group_prioritization_tbl %>% 
  filter(fraction_ligand_group > fraction_cutoff & fraction_receptor_group > fraction_cutoff) %>% 
  distinct(id, sender, receiver, ligand, receptor, group, prioritization_score, ligand_receptor_lfc_avg, fraction_expressing_ligand_receptor) %>% 
  filter(ligand_receptor_lfc_avg > 0 & fraction_expressing_ligand_receptor > 0) %>% top_n(100, prioritization_score) 

prioritized_tbl_oi = output$prioritization_tables$group_prioritization_tbl %>% 
  filter(id %in% prioritized_tbl_oi$id) %>% 
  distinct(id, sender, receiver, ligand, receptor, group) %>% left_join(prioritized_tbl_oi)
prioritized_tbl_oi$prioritization_score[is.na(prioritized_tbl_oi$prioritization_score)] = 0


n_senders = prioritized_tbl_oi$sender %>% unique() %>% length()
n_receivers = prioritized_tbl_oi$receiver %>% unique() %>% length()

  
colors_sender = c("red","tomato","violetred3", "mediumpurple1", "royalblue")
colors_receiver = c("darkolivegreen1", "darkolivegreen2", "darkolivegreen3", "lawngreen", "limegreen")

circos_list = make_circos_group_comparison(prioritized_tbl_oi, colors_sender, colors_receiver)

```

## Visualization of ligand-activity, ligand-target links, and target gene expression

```{r}
group_oi = "High"
receiver_oi = "Myeloid"
prioritized_tbl_oi = output$prioritization_tables$group_prioritization_tbl %>% 
  filter(fraction_ligand_group > fraction_cutoff & fraction_receptor_group > fraction_cutoff) %>% 
  distinct(id, sender, receiver, ligand, receptor, group, prioritization_score, ligand_receptor_lfc_avg, fraction_expressing_ligand_receptor, activity_scaled) %>% 
  filter(ligand_receptor_lfc_avg > 0 & fraction_expressing_ligand_receptor > 0) %>% 
  filter(group == group_oi & receiver == receiver_oi) %>% 
  group_by(group) %>% top_n(50, prioritization_score) %>% top_n(25, activity_scaled) %>% arrange(-activity_scaled)
```

```{r, fig.width=18, fig.height=10}
combined_plot = make_ligand_activity_target_plot(group_oi, receiver_oi, prioritized_tbl_oi, output$ligand_activities_targets_DEgenes, contrast_tbl, output$grouping_tbl, output$celltype_info, plot_legend = FALSE)
combined_plot
```
## Show ligand activities for each receiver-group combination



Both scaled and absolute activities.

In combination with ligand-receptor expression per sample


```{r}
group_oi = "High"
```

```{r, fig.width=20, fig.height=22}

prioritized_tbl_oi = output$prioritization_tables$group_prioritization_tbl %>% 
  filter(fraction_ligand_group > fraction_cutoff & fraction_receptor_group > fraction_cutoff) %>% 
  distinct(id, sender, receiver, lr_interaction, group, ligand_receptor_lfc_avg, activity_scaled, fraction_ligand_group, fraction_expressing_ligand_receptor, scaled_avg_exprs_ligand, prioritization_score) %>% 
  filter(ligand_receptor_lfc_avg > 0 & fraction_expressing_ligand_receptor > 0) %>% 
  filter(group == group_oi)  %>% top_n(250, prioritization_score) %>% group_by(receiver) %>% top_n(10)

plot_oi = make_sample_lr_prod_activity_plots(output$prioritization_tables,prioritized_tbl_oi, widths = c(8,2,2))
plot_oi
plot_oi
```

Both scaled and absolute activities.

Without ligand-receptor expression per sample

Show top ligands based on ligand activity (agnostic of expression in sender)

```{r, fig.width=5, fig.height=8}

ligands_oi = output$prioritization_tables$ligand_activities_target_de_tbl %>% inner_join(contrast_tbl) %>% 
  group_by(group, receiver) %>% distinct(ligand, receiver, group, activity) %>% 
  top_n(5, activity) %>% pull(ligand) %>% unique()

plot_oi = make_ligand_activity_plots(output$prioritization_tables, ligands_oi, contrast_tbl, widths = NULL)
plot_oi

```
Show top ligands based on prioritization scores

```{r, fig.width=5, fig.height=8}

ligands_oi = output$prioritization_tables$group_prioritization_tbl %>% 
  group_by(group, receiver) %>% distinct(ligand, receiver, group, prioritization_score) %>% 
  top_n(5, prioritization_score) %>% pull(ligand) %>% unique()

plot_oi = make_ligand_activity_plots(output$prioritization_tables, ligands_oi, contrast_tbl, widths = NULL)
plot_oi

```
